### ISSUE TITLE
Suggest unified reward function/model architecture for a wide range of tasks

### ISSUE THREAD

**Issue by James4Ever0** [lucidrains/PaLM-rlhf-pytorch#4]
I find the reward function to be the most important part of RLHF, because it is the part which mimics a human evaluator, providing instant feedback to the model.

However, due to ChatGPT's wide range of language capabilities, it is hard to model such reward function with a single model to be prompt dependent, context aware, leveraging existing knowledge from pretrained models.

Most projects relating to RLHF usually use toy-like reward function such as counting word frequencies, checking output formats, or just sentiment/fluency scores. These functions do not "think" like the human evaluator, considering every factor as a whole. RL4LMs propose [GRUE](https://rl4lms.apps.allenai.org/grue) in which the model performs general instructions but it does not expose a simple unified interface to get a score given prompt and answer.

RL4LMs contains a [registry](https://github.com/allenai/RL4LMs/blob/1c0fe2f4071bc8b9522ebc0dd15e6207d9dfc4f8/rl4lms/envs/text_generation/registry.py#L65) of [reward functions](https://github.com/allenai/RL4LMs/blob/1c0fe2f4071bc8b9522ebc0dd15e6207d9dfc4f8/rl4lms/envs/text_generation/reward.py#L390) which I find it complex and not leveraging existing pretrained models. I think a reward function should be an integrated part of the language model itself, rather than outsourcing it to different architectures, able to attribute the reward to fine-grained sections of outputs.

**Comment by James4Ever0** [comment]
RLHF requires creating multiple models like SFT, RM, PPO-tuned model. Is it possible to increase storage and memory efficiency, reduce computation if we freeze some huge layers of the pretrained model, only fine-tune certain layers to create SFT, RM, PPO using [OpenDelta](https://opendelta.readthedocs.io/en/latest) or other libraries/methods? I read that your repo is using LoRA but I'm not sure if it fulfills all goals described above. Common implementations like [minRLHF](https://github.com/thomfoster/minRLHF/blob/main/examples/huggingface_example.ipynb) requires four separate models, three are derived from the pretrained model as actor, critic and reference, in addition to an external sentiment rating model.

**Comment by James4Ever0** [comment]
To address this proposal even further, I think a good reward function can self-evolve and adapt to new environments, making this model multipurpose, realtime and even into AGI.
